### PR TITLE
Unify IOperation APIs for while and do loops into a single IWhileLoop…

### LINKED ISSUE #22960: Unify IOperation API for do an while loop statements
This was concluded in last design meeting but we did not have a tracking issue.

### PR DESCRIPTION
**Pull request by mavasani**
…Operation

Fixes #22960
https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/518814

### CODE REVIEW

**Comment by 333fred** [comment]
>     public interface IWhileLoopOperation : ILoopOperation

Feels like this should have a different name now that it's for both do and while loops. Perhaps `ILoopOperation`, although that has conflicts with for loops.

---
Refers to: src/Compilers/Core/Portable/Operations/Loops/IWhileLoopOperation.cs:17 in 4dd2a44. [](commit_id = 4dd2a449b0031de40af98c032547ae12dc55811e, deletion_comment = False)

**Comment by 333fred** [comment]
>     public interface IWhileLoopOperation : ILoopOperation

Though they both do have while in the syntax, so I suppose it's not a big concern.

---
In reply to: [341500157](https://github.com/dotnet/roslyn/pull/22965#issuecomment-341500157) [](ancestors = 341500157)

---
Refers to: src/Compilers/Core/Portable/Operations/Loops/IWhileLoopOperation.cs:17 in 4dd2a44. [](commit_id = 4dd2a449b0031de40af98c032547ae12dc55811e, deletion_comment = False)

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 1]
```diff
@@ -436,25 +436,15 @@ public override void VisitSwitchCase(ISwitchCaseOperation operation)
...
-
         public override void VisitWhileLoop(IWhileLoopOperation operation)
         {
             LogString(nameof(IWhileLoopOperation));
+            LogString($" (ConditionIsTop: {operation.ConditionIsTop}, ConditionIsUntil: {operation.ConditionIsUntil})");
             LogLoopStatementHeader(operation);
 
             Visit(operation.Condition, "Condition");
             Visit(operation.Body, "Body");
+            Visit(operation.IgnoredCondition, "IgnoredCondition");
```
Comment (333fred) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: Why not print this as DoLoop did, so we minimize diff?

**Comment by mavasani** [comment]
@dotnet-bot retest ubuntu_16_debug_prtest please

**Comment by mavasani** [comment]
> Feels like this should have a different name now that it's for both do and while loops. Perhaps ILoopOperation, although that has conflicts with for loops.

We had discussed the naming at the design meeting. ILoopOperation is already base interface for all loops and this loop semantically represents a condition based loop. `IWhileLoopOperation` seemed most intuitive choice for the name.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 1 above)
Comment (mavasani) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: DoLoop also always printed the IgnoreCondition even when null, more diffs are due to the fact that now even while loops have this field - I think we decided on printing null members in the dumper for more clarity some time ago.

**Review by 333fred — APPROVED** [review]
LGTM. Still think we should change the order of condition printing, but that's more of a nit.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 1 above)
Comment (333fred) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: There are more than 5 diffs of just moving the ignored condition. It's not a big deal though.

**Comment by natidea** [comment]
Approved to merge via [518814](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/518814)

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler Can I get code reviews please?

**Review by AlekseyTs — APPROVED** [review]
LGTM

**Comment by mavasani** [comment]
Can I get one more review from @dotnet/roslyn-compiler please? This is a 15.5 fix for IOperation.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 2]
```diff
@@ -1309,26 +1309,30 @@ private IWhileLoopOperation CreateBoundWhileStatementOperation(BoundWhileStateme
         {
             Lazy<IOperation> condition = new Lazy<IOperation>(() => Create(boundWhileStatement.Condition));
             Lazy<IOperation> body = new Lazy<IOperation>(() => Create(boundWhileStatement.Body));
+            Lazy<IOperation> ignoredCondition = new Lazy<IOperation>(() => null);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Is there a singleton instance that can be used here?

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 3]
```diff
@@ -1237,12 +1215,15 @@ Namespace Microsoft.CodeAnalysis.Operations
         Private Function CreateBoundWhileStatementOperation(boundWhileStatement As BoundWhileStatement) As IWhileLoopOperation
             Dim condition As Lazy(Of IOperation) = New Lazy(Of IOperation)(Function() Create(boundWhileStatement.Condition))
             Dim body As Lazy(Of IOperation) = New Lazy(Of IOperation)(Function() Create(boundWhileStatement.Body))
+            Dim ignoredCondition As Lazy(Of IOperation) = New Lazy(Of IOperation)(Function() Nothing)
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Singleton?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (mavasani) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: That would create duplicate operation nodes in the tree.

Comment (cston) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: This `Lazy<IOperation>` generates a `null` instance.

**Comment by mavasani** [comment]
@dotnet-bot retest ubuntu_16_debug_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Pull request merged**